Graphitey consolidation part 1

Dieterbe opened the pull request.

Dieterbe commented:
@woodsaj or @replay if you have some time, mind taking a look? this covers step 1 of https://github.com/raintank/metrictank/issues/463#issuecomment-279966340

replay commented on metrictank.go:
```diff
@@ -417,6 +392,10 @@ func main() {
 	// When the timer becomes 0 it means the in-memory buffer has been able to fully populate so that if you stop a primary
 	// and it was able to save its complete chunks, this node will be able to take over without dataloss.
 	// You can upgrade a candidate to primary while the timer is not 0 yet, it just means it may have missing data in the chunks that it will save.
+	highestChunkSpan := chunkSpan
```
^^ That's between the comment on line `391 - 394` and the code the comment is talking about

replay commented on api/query_engine.go:
```diff
@@ -73,15 +95,120 @@ func getOptions(aggSettings mdata.AggSettings, minInterval, tsRange uint32) ([]a
+	}
+	lcm := util.Lcm(keys)
+
+	options := make([]narchive, 0, len(aggSettings.Aggs)+1)
```
instead of `[]narchive` you could just use `narchives`

replay commented on api/query_engine.go:
```diff
@@ -73,15 +95,120 @@ func getOptions(aggSettings mdata.AggSettings, minInterval, tsRange uint32) ([]a
+// potentially making the "raw interval" > first rollup interval, so that
+// raw archive might be ordered after a rollup, and where the archive
+// itself knows which one it is, instead of having to rely on index lookup table.
+func getOptionsLikeGraphite(reqs []models.Req, aggSettings mdata.AggSettings, tsRange uint32) ([]narchive, map[uint32]struct{}) {
```
could just use `narchives` instead of `[]narchive` if the type is already defined anyway

replay commented on api/query_engine.go:
```diff
@@ -73,15 +95,120 @@ func getOptions(aggSettings mdata.AggSettings, minInterval, tsRange uint32) ([]a
+	}
+
+	if LogLevel < 2 {
+		options[selected].chosen = true
```
Even though this saves an operation in prod, I wouldn't modify such an important data structure under a condition that depends on the loglevel even if at the moment it isn't used at any other critical location (not considering `narchive.String()` critical) because I think it's too easy for somebody to use that `.chosen` flag for something else without being aware that it's correctness depends on such an arbitrary setting like the log level.

replay commented on api/query_engine.go:
```diff
@@ -37,6 +38,27 @@ func (a archives) Len() int           { return len(a) }
+// number 0 = raw archive, number 1,2,..N = aggregation 1,2..N
+// this allows you to reorder narchives and still keep track of which is which
+type narchive struct {
+	i          int
```
To understand the purpose of this `i` it's necessary to go quite deep into how `narchive` is used. Some explanatory name or some comment why it is necessary to keep that instead of just relying on the index of the `options` list would be helpful.

replay commented:
Looks good, with some minor comments

woodsaj commented:
This seems to be way more complicated then it should be.  We are simplifying our code, so there should be more lines of code deleted then new lines being added.

1) i dont like the idea of supporting 2 query modes. We are building a graphite compatible backend, so  there is no need for a "make incompatible" flag.  Supporting the 2 modes just complicates the code.

2) The logic for selecting the archive to use for a query seems convoluted.  all we need to do is return the highest resolution archive that has data for the requested time range.  As all series have the same aggSettings, and aggSettings.Aggs should already be sorted by data resolution (if we dont re-order when parsing the config, we should)
```
func archiveForRequest( aggSettings mdata.AggSettings, from int32) int {
  now := time.Now().Unix()
  if now-aggSettings.RawTTL <= from {
    return 0  // should probably use a constant like mdata.RawArchive
  }
  for i, aggSetting := range aggSettings.Aggs {
    if now-agg.TTL <= from {
       return i+1
   }
  }
  return len(aggSetting.Aggs)
}
```

Dieterbe commented:
> i dont like the idea of supporting 2 query modes. We are building a graphite compatible backend, so there is no need for a "make incompatible" flag. Supporting the 2 modes just complicates the code.

I think there should be a grace period between introducing the new code and getting rid of the old. 
We should only get rid of the old code once we've given the new approach some time to mature (and possibly make adjustments) and prove that it works well for all of our environments.
There's no harm in keeping the old code around for a bit longer, and it gives us more options. (e.g. in case it doesn't work well yet for an environment but we'd like to update MT because we want to roll out other things)

> if we dont re-order when parsing the config, we should

no. user should specify them in the right order.  it's for their own good (see also the commit msg of https://github.com/raintank/metrictank/pull/534/commits/6dc62be95d3c9539109b64f00d4d59fa94ac9e2b)

> The logic for selecting the archive to use for a query seems convoluted. all we need to do is return the highest resolution archive that has data for the requested time range

no. we also need to deal with multiple different raw intervals (and how their lcm may be > 1st rollup) and with the per-aggsetting ready parameter. I think I can replace the ready filtering in `getOptionsLikeGraphite` in favor of just checking the ready state in `alignRequestsLikeGraphite`. this should save about 5 lines.  Can you point to any specific items that you think can be removed?

woodsaj commented:
> I think there should be a grace period between introducing the new code and getting rid of the old.

the old code is broken and returns incorrect data, get rid of it.

> no. we also need to deal with multiple different raw intervals (and how their lcm may be > 1st rollup)

that is not how graphite handles queries. The only information used for selecting the rollup is the request "from" time.  Users sending data with an interval less frequent then the first rollup is an edge case we dont need to support right now.  that will be solved by supporting storeage-schema config (ie per series aggSettings)

woodsaj commented on api/query_engine.go:
```diff
@@ -38,41 +38,178 @@ func (a archives) Len() int           { return len(a) }
+// represents a numbered data "archive", i.e. the raw one, or an aggregated series
+// number 0 = raw archive, number 1,2,..N = aggregation 1,2..N
+// this allows you to reorder narchives and still keep track of which is which
+type narchive struct {
```
what does narchive mean?   In general just dont use abbrivations.

woodsaj commented on api/query_engine.go:
```diff
@@ -38,41 +38,178 @@ func (a archives) Len() int           { return len(a) }
+type narchive struct {
+	i          int
+	interval   uint32
+	pointCount uint32
```
this is not used anywhere except in the String() method

woodsaj commented on api/query_engine.go:
```diff
@@ -38,41 +38,178 @@ func (a archives) Len() int           { return len(a) }
-// note: it is assumed that all requests have the same from, to and maxdatapoints!
-// this function ignores the TTL values. it is assumed that you've set sensible TTL's
-func alignRequests(reqs []models.Req, aggSettings mdata.AggSettings) ([]models.Req, error) {
+// represents a numbered data "archive", i.e. the raw one, or an aggregated series
```
Why are there 2 almost identical data structures (archive and narchive).  Surely you can just add the position variable (i) to archive and remove narchive.  Just because it is defined doesnt mean it has to be used.

Dieterbe commented:
I removed the old approach as requested, and simplified more. (see last commit above)
been testing/experimenting a bit. on my pc going to 500k~1M points makes it take several seconds. MT reports upto 80ms. seems several seconds are spent in graphite and/or sending data to graphite. up to 200k points is pleasant (fraction of a second response when no processing in graphite). I think we might want to shorten the TTL of raw data for some of our customers to avoid this being an issue. (we have some customers where we maintain raw 35 days and first rollup 38days)

Dieterbe commented:
@woodsaj PTAL and merge if you like.

woodsaj commented:
looks great @Dieterbe 

We can likely reduce the query times by enabling gzip compression on the data sent between MT and graphite. https://go-macaron.com/docs/middlewares/gzip  Ill open a separate issue for that.
